crates/cards/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+            Ok(Self(secret))
+        } else {
+            Err(report!(errors::ValidationError::InvalidValue {
+                message: "invalid card security code".to_string()
+            }))
+        }
+    }
+}
+
+pub struct CardExpirationMonth(StrongSecret<u8>);
+
+impl CardExpirationMonth {
+    pub fn new(secret: StrongSecret<u8>) -> errors::CustomResult<Self, errors::ValidationError> {
+        let month = secret.peek();
+
+        if *month >= 1 && *month <= 12 {
+            Ok(Self(secret))
+        } else {
+            Err(report!(errors::ValidationError::InvalidValue {
+                message: "invalid card expiration month".to_string()
+            }))
+        }
+    }
+
+    pub fn two_digits(&self) -> String {
+        let month = self.0.peek();
+        let month_str: String = month.to_string();
+        if month_str.len() == 1 {
+            format!("0{}", month_str)
+        } else {
+            month_str
+        }
+    }
+}
+
+pub struct CardExpirationYear(StrongSecret<u16>);
+
+impl CardExpirationYear {
+    pub fn new(secret: StrongSecret<u16>) -> errors::CustomResult<Self, errors::ValidationError> {
+        let year = secret.peek();
+
+        if *year >= 1997 {
+            Ok(Self(secret))
+        } else {
+            Err(report!(errors::ValidationError::InvalidValue {
+                message: "invalid card expiration year".to_string()
+            }))
+        }
+    }
+
+    pub fn four_digits(&self) -> String {
+        self.0.peek().to_string()
+    }
+
+    pub fn two_digits(&self) -> String {
+        let year = &self.0.peek().to_string()[2..4];
+        year.to_string()
+    }
+}
+
+pub struct CardExpiration {
+    pub month: CardExpirationMonth,
+    pub year: CardExpirationYear,
+}
+
+impl CardExpiration {
+    pub fn new(
+        secret_month: StrongSecret<u8>,
+        secret_year: StrongSecret<u16>,
+    ) -> errors::CustomResult<Self, errors::ValidationError> {
+        let card_month = CardExpirationMonth::new(secret_month);
+
+        match card_month {
+            Ok(cm) => {
+                let card_year = CardExpirationYear::new(secret_year);
+
+                match card_year {
+                    Ok(cy) => Ok(Self {
+                        month: cm,
+                        year: cy,
+                    }),
+                    Err(e) => Err(e),
+                }
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn is_expired(&self) -> bool {
+        let current_datetime_utc = date_time::now();
+
+        // card expiry day is last day of the expiration month
+        let expiration_day = days_in_year_month(
+            i32::from(*self.year.0.peek()),
+            (*self.month.0.peek()).try_into().unwrap(),
+        );
+
+        let expiration_date = Date::from_calendar_date(
+            i32::from(*self.year.0.peek()),
+            (*self.month.0.peek()).try_into().unwrap(),
+            expiration_day,
+        )
+        .unwrap();
```

Comment:
We preferably don't want to panic for invalid values.  

@NishantJoshi00 @vspecky Should we have this function return `Result<bool, std::convert::Infallible>`?

Reply:
I see 2 approaches in which we can handle this:
1. We can how it return `Option<boo>` and then using `.ok()` to convert those `Result`s into `Option`s.
2. We can do the same thing but instead of return type being an option we can keep it as bool and do `unwrap_or(false)`

Reply:
I'd prefer returning a `Result<bool, SomeError>` here. We cannot just unwrap the result since in certain cases, the date may be invalid (eg: `29th Feb 2025` which is invalid cuz 2025 is not a leap year)

That, or we perform this validation during construction (makes more sense) in the `new()` function. Then, given that all members are private, it should be fine to `.expect()` here.

cc: @NishantJoshi00

Reply:
Okay so apparently the `time` crate handles cases like the one I mentioned above. In that case, I'm fine with an `.expect()` here. We can add a `time` sanity test that ensures any crate version bump does not break panic safety.

Also, for every `.expect()`, include a comment
```
// Panic Safety: reason why this will not panic
```

Reply:
@vspecky @NishantJoshi00 can you check me if below implementation is as suggested 

```rust
pub fn is_expired(&self) -> bool {
        let current_datetime_utc = date_time::now();
        
        // Panic Safety: struct CardExpirationMonth validates month  
        let expiration_month = (*self.month.0.peek()).try_into().expect("valid month in range 1 to 12");

        let expiration_year = *self.year.0.peek();

        // card expiry day is last day of the expiration month
        let expiration_day = days_in_year_month(i32::from(expiration_year), expiration_month);

        // Panic Safety: year, month, day validated
        let expiration_date =
            Date::from_calendar_date(i32::from(expiration_year), expiration_month, expiration_day)
                .expect("valid date");

        let expiration_time = Time::MIDNIGHT;

        // actual expiry date specified on card w.r.t. local timezone
        // max diff b/w utc and other timezones is 14 hours
        let mut expiration_datetime_utc = PrimitiveDateTime::new(expiration_date, expiration_time);

        // compensating time difference b/w local and utc timezone by adding a day
        expiration_datetime_utc = expiration_datetime_utc.saturating_add(Duration::days(1));

        if current_datetime_utc > expiration_datetime_utc {
            true
        } else {
            false
        }
    }

```

Reply:
@SanchithHegde @vspecky @NishantJoshi00 thoughts on this ?